### DISCUSSION TITLE
How to have multiple filters for same column

### DISCUSSION THREAD

**Discussion by jaekunchoi** [TanStack/table#2290] [General]
Is there a way to implement multiple filters for one column?

I want to be able to use word column and have various of filters for that. Currently there is only one filter I am using with `Filter` column option which is working fine but I can't find an example of using multiple filters for this same column for example a word column will need `and/or` filter as well as multiple words as input value for filtering rows.

**Comment by brianhankins** [comment]
I'm also looking for a solution to this. The current app I'm working on is using v6 branch. 

We've tried using this: [React-Table - Custom Filtering + Multiple values filte](https://codesandbox.io/s/3x51yzollq?file=/index.js)

However, it seems that when passing in an array to the onFilteredChangeCustom(value, accessor) method, the table wont display any rows if the array contains more than one element. 

This was also asked 7 months ago on SO: [react-table-custom-fitering-doesnt-work-with-multiselect-react-select](https://stackoverflow.com/questions/58187252/react-table-custom-fitering-doesnt-work-with-multiselect-react-select)

It currently doesn't have any answers. I'm not sure when this broke or how the developer using the code sandbox got this to work. I also noticed though in the code sandbox that you have to use the react-select. Manually typing into the filter on the table wont produce an array and will show no rows.  

I rolled back my version to React-Table: 6.5.3 as shown in the code sandbox and still this isn't working either. I'm not sure how this code sandbox is working...

**Reply by danish1994** [reply]
In the above example - https://codesandbox.io/s/react-table-custom-filtering-multiple-values-filter-forked-1f98i
If you see `defaultFilterMethod` is being added. which is resolving an array of values.

**Comment by ogmzh** [comment]
I'm aware you asked this in 2020, but this is how I did it in v7:
When creating a column level filter, you can pass an array as the last argument (which is usually called `filterValue` as a single string), and the column filter looks like:
```
return {
            Header: label,
            accessor: key,
            filter: (rows: Row<MyEntity>[], columnIds: string[], filterValues: string[]) => {
              if (!filterValues || filterValues.length === 0) {
                return rows;
              }
              return rows.filter(row => filterValues.includes(row.values[columnIds[0]]));
            },
          };
```

Adding values to this filter, i flatmap new values into a single array (which is then picked up by the `filterValues` array mentioned above:
 ```
         setFilter(columnId, [
                      ...filters
                        .filter(existingFilter => existingFilter.id === columnId)
                        .flatMap(existingFilter => existingFilter.value),
                      newFilterValue,
                    ]);
```

**Reply by joonasstenvall** [reply]
Hi, I know that this is a late repy but could you explain this a bit more? I'm having a hard time with this myself. I need to be able to filter a column atleast twice. Eg, Select names foo and bar and columns would only show data from foo and bar.

**Reply by joonasstenvall** [reply]
Just as I replied I stumbled across this thread which solved my issue!
https://github.com/tannerlinsley/react-table/discussions/2350
Leaving this for future googlers!

**Comment by phb-ig** [comment]
This no longer works in v8. [The column filterFn only receives the current row, the current filter ID, and a single filterValue](https://tanstack.com/table/v8/docs/guide/column-filtering#custom-filter-functions). All filters specified in the `columnFilters` array are run, but only the one which comes sequentially last is applied.

One workaround would be to add a custom `filterFn` that checks to see if an array filter value contains the column's value, but in order to apply these filters correctly, you would also need to grab the item for the column being filtered out of `columnFilters` and edit it, rather than just adding/removing a new item for each filter.

**Comment by timakro** [comment]
This is TanStack Table's column filter state:

```ts
interface ColumnFilter {
  id: string
  value: unknown
}
type ColumnFiltersState = ColumnFilter[]
```

Since `value` has type `unknown` you can put anything there. So one option is to use an array of strings as the filter value and define your own `filterFn` to handle that:

```ts
function filterFn<TData>(row: Row<TData>, columnId: string, filterValue: string[]) {
  const rowValue = row.getValue(columnId)
  return filterValue.some((v) => String(rowValue).toLowerCase().includes(v.toLowerCase())
}
```

As @phb-ig pointed out that makes it a little bit more tricky to add a filter to a column that already has a filter. So if you want to actually have multiple `ColumnFilter` with the same `id` (which is what I wanted) you can do the following:

```ts
export function columnFilterFn<TData>(row: Row<TData>, columnId: string, filterValue: ColumnFilterValue[]) {
  // If previous filter on the same column has marked this row as excluded, keep it excluded regardless of current filter value.
  if (row.columnFilters[columnId] === false) {
    return false
  }

  const rowValue = row.getValue(columnId)
  return filterValue.some((v) => String(rowValue).toLowerCase().includes(v.toLowerCase())
}
```

Note that I am combining both approaches to where items in the outer list are semantically ANDed and items in the inner list are semantically ORed. These semantics are common in many filtering REST APIs to which you might want to connect your table, e.g. `?color=blue&size=m,s`
